Source/qol/stash.cpp

```diff
@@ -0,0 +1,609 @@
+{
+	InitialWithdrawGoldValue = Stash.gold;
+
+	if (talkflag)
+		control_reset_talk();
+
+	Point start = GetPanelPosition(UiPanels::Stash, { 67, 128 });
+	SDL_Rect rect = MakeSdlRect(start.x, start.y, 180, 20);
+	SDL_SetTextInputRect(&rect);
+
+	IsWithdrawGoldOpen = true;
+	WithdrawGoldValue = 0;
+	SDL_StartTextInput();
+}
+
+void WithdrawGold(Player &player, int amount)
+{
+	InitializeItem(player.HoldItem, IDI_GOLD);
+	SetGoldSeed(player, player.HoldItem);
+	player.HoldItem._ivalue = amount;
+	player.HoldItem._iStatFlag = true;
+	ControlSetGoldCurs(player);
+	Stash.gold -= amount;
+	Stash.dirty = true;
+}
+
+} // namespace
+
+Point GetStashSlotCoord(Point slot)
+{
+	constexpr int StashNextCell = INV_SLOT_SIZE_PX + 1; // spacing between each cell
+
+	return GetPanelPosition(UiPanels::Stash, slot * StashNextCell + Displacement { 17, 48 });
+}
+
+void FreeStashGFX()
+{
+	StashPanelArt.Unload();
+	StashNavButtonArt.Unload();
+}
+
+void InitStash()
+{
+	InitialWithdrawGoldValue = 0;
+
+	LoadArt("data\\stash.pcx", &StashPanelArt, 1);
+	LoadArt("data\\stashnavbtns.pcx", &StashNavButtonArt, 5);
+}
+
+int StashButtonPressed = -1;
+
+void CheckStashButtonRelease(Point mousePosition)
+{
+	if (StashButtonPressed == -1)
+		return;
+
+	Rectangle stashButton = StashButtonRect[StashButtonPressed];
+	stashButton.position = GetPanelPosition(UiPanels::Stash, stashButton.position);
+	if (stashButton.Contains(mousePosition)) {
+		switch (StashButtonPressed) {
+		case 0:
+			Stash.page -= 10;
+			break;
+		case 1:
+			Stash.page -= 1;
+			break;
+		case 2:
+			StartGoldWithdraw();
+			break;
+		case 3:
+			Stash.page += 1;
+			break;
+		case 4:
+			Stash.page += 10;
+			break;
+		}
+	}
+
+	StashButtonPressed = -1;
+	Stash.page = clamp(Stash.page, 0, 49);
+}
+
+void CheckStashButtonPress(Point mousePosition)
+{
+	Rectangle stashButton;
+
+	for (int i = 0; i < 5; i++) {
+		stashButton = StashButtonRect[i];
+		stashButton.position = GetPanelPosition(UiPanels::Stash, stashButton.position);
+		if (stashButton.Contains(mousePosition)) {
+			StashButtonPressed = i;
+			return;
+		}
+	}
+
+	StashButtonPressed = -1;
+}
+
+void DrawStash(const Surface &out)
+{
+	DrawArt(out, GetPanelPosition(UiPanels::Stash), &StashPanelArt);
+
+	if (StashButtonPressed != -1) {
+		Point stashButton = GetPanelPosition(UiPanels::Stash, StashButtonRect[StashButtonPressed].position);
+		DrawArt(out, stashButton, &StashNavButtonArt, StashButtonPressed);
+	}
+
+	constexpr Displacement offset { 0, INV_SLOT_SIZE_PX - 1 };
+
+	std::vector<uint16_t> itemIds = {};
+
+	for (auto slot : PointsInRectangleRange({ { 0, 0 }, { 10, 10 } })) {
+		if (Stash.stashGrids[Stash.page][slot.x][slot.y] == 0) {
+			continue; // Not item in the given slot
+		}
+
+		InvDrawSlotBack(out, GetStashSlotCoord(slot) + offset, InventorySlotSizeInPixels);
+
+		uint16_t itemId = Stash.stashGrids[Stash.page][slot.x][slot.y] - 1;
+		if (Stash.stashList[itemId].position == slot) {
+			itemIds.push_back(itemId);

```

Comment:
This `itemIds` vector is only used once below. Rather than creating a vector on every frame, simply move the body of the loop below in here.

Reply:
The background must first be drawn before the items are drawn on top. The vector is to save us from iterating over all slots twice and instead remember which ones contains items from the first loop.

Reply:
Ah ok, though I bet scanning twice is faster than allocating.

Reply:
Alright, I can revert it :)